src/interfaces/object.rs

```diff
@@ -1,77 +1,70 @@
-use crate::foundation::{IReference, PropertyValue};
+use crate::foundation::{IReference, IStringable, PropertyValue};
 use crate::*;
 
 /// A WinRT object that may be used as a polymorphic stand-in for any WinRT class, interface, or boxed value.
-///
-/// Objects implement the [IInspectable](https://docs.microsoft.com/en-us/windows/win32/api/inspectable/nn-inspectable-iinspectable) interface.
+/// `Object` implements the
+/// [IInspectable](https://docs.microsoft.com/en-us/windows/win32/api/inspectable/nn-inspectable-iinspectable)
+/// interface.
 #[repr(transparent)]
-#[derive(Default, Clone)]
-pub struct Object {
-    ptr: ComPtr<Object>,
-}
+#[derive(Clone, PartialEq)]
+pub struct Object(IUnknown);
 
 impl Object {
-    /// Returns the fully-qualified type name of the WinRT object.
+    /// Returns the canonical type name for the underlying object.
     pub fn type_name(&self) -> Result<HString> {
-        let this = self
-            .get_abi()
-            .expect("The `this` pointer was null when calling method");
-
-        let mut string = HString::default();
-
         unsafe {
-            (this.vtable().inspectable_type_name)(this, string.set_abi()).ok()?;
+            let mut abi = std::ptr::null_mut();
+            (self.vtable().4)(self.abi(), &mut abi).ok()?;
+            Ok(std::mem::transmute(abi))
         }
-
-        Ok(string)
     }
 }
 
-unsafe impl ComInterface for Object {
-    type VTable = abi_IInspectable;
-
-    const IID: Guid = {
-        Guid::from_values(
-            0xAF86_E2E0,
-            0xB12D,
-            0x4C6A,
-            [0x9C, 0x5A, 0xD7, 0xAA, 0x65, 0x10, 0x1E, 0x90],
-        )
-    };
+#[repr(C)]
+pub struct Object_vtable(
```

Comment:
It was nicer IMO when these structs had field names so you could refer to the function pointers by name instead of by offset.

Reply:
There are pros and cons. Naming them helps a bit with readability on the consumption side but makes the code for implementing the interface much more complicated to generate and read. Names are also problematic because we have to do a lot of renaming and mangling to avoid collisions and then figuring out what the original names were is a little more complicated.